diff -up openhpi-3.2.1/Makefile.am.docs openhpi-3.2.1/Makefile.am
--- openhpi-3.2.1/Makefile.am.docs	2013-08-20 18:20:30.498731238 +0200
+++ openhpi-3.2.1/Makefile.am	2013-08-20 18:20:43.124971660 +0200
@@ -116,8 +116,7 @@ include_HEADERS = $(top_srcdir)/include/
                   $(top_srcdir)/utils/event_utils.h \
                   $(top_srcdir)/clients/oh_clients.h
 
-docsdir=$(datadir)/doc/$(PACKAGE_NAME)-$(VERSION)
-docs_DATA = README README.daemon COPYING
+doc_DATA = README README.daemon COPYING
 
 $(top_srcdir)/utils/sahpi_enum_utils.h:
 	make -C $(top_builddir)/utils sahpi_enum_utils.h
