diff -up openhpi-2.14.1/plugins/oa_soap/oa_soap_resources.c.thermal-info-segfault openhpi-2.14.1/plugins/oa_soap/oa_soap_resources.c
--- openhpi-2.14.1/plugins/oa_soap/oa_soap_resources.c.thermal-info-segfault	2013-04-03 22:56:35.856017826 +0200
+++ openhpi-2.14.1/plugins/oa_soap/oa_soap_resources.c	2013-04-03 23:26:49.296892952 +0200
@@ -9928,10 +9928,13 @@ const char *oa_soap_bld_type_str[] = {
 	"BL460C G8",
 	"BL460C G7",
 	"BL460C",
+	"BL465C G7",
 	"BL465C",
+	"BL480C G1",
 	"BL480C",
 	"BL495C",
 	"BL680C",
+	"BL685C G6",
 	"BL685C",
 	"BL860C",
 	"BL870C",
@@ -10034,6 +10037,18 @@ const struct oa_soap_static_thermal_sens
 		{OA_SOAP_SEN_BLADE_MEM_ZONE1, MEMORY_ZONE, 1},
 		{OA_SOAP_SEN_TEMP_STATUS, AMBIENT_ZONE, 1}
 	},
+	/* BL465c G7 blade type */
+	{
+		{OA_SOAP_SEN_BLADE_SYSTEM_ZONE1, SYSTEM_ZONE, 0},
+		{OA_SOAP_SEN_BLADE_CPU_ZONE1, CPU_ZONE, 1},
+		{OA_SOAP_SEN_BLADE_CPU1_1, CPU_1, 1},
+		{OA_SOAP_SEN_BLADE_CPU2_1, CPU_2, 1},
+		{OA_SOAP_SEN_BLADE_CPU3_1, CPU_3, 0},
+		{OA_SOAP_SEN_BLADE_CPU4_1, CPU_4, 0},
+		{OA_SOAP_SEN_BLADE_DISK_ZONE1, DISK_ZONE, 0},
+		{OA_SOAP_SEN_BLADE_MEM_ZONE1, MEMORY_ZONE, 2},
+		{OA_SOAP_SEN_TEMP_STATUS, AMBIENT_ZONE, 1}
+	},
 	/* BL465c blade type */
 	{
 		{OA_SOAP_SEN_BLADE_SYSTEM_ZONE1, SYSTEM_ZONE, 1},
@@ -10046,6 +10061,18 @@ const struct oa_soap_static_thermal_sens
 		{OA_SOAP_SEN_BLADE_MEM_ZONE1, MEMORY_ZONE, 2},
 		{OA_SOAP_SEN_TEMP_STATUS, AMBIENT_ZONE, 1}
 	},
+	/* BL480c G1 blade type */
+	{
+		{OA_SOAP_SEN_BLADE_SYSTEM_ZONE1, SYSTEM_ZONE, 0},
+		{OA_SOAP_SEN_BLADE_CPU_ZONE1, CPU_ZONE, 1},
+		{OA_SOAP_SEN_BLADE_CPU1_1, CPU_1, 2},
+		{OA_SOAP_SEN_BLADE_CPU2_1, CPU_2, 2},
+		{OA_SOAP_SEN_BLADE_CPU3_1, CPU_3, 0},
+		{OA_SOAP_SEN_BLADE_CPU4_1, CPU_4, 0},
+		{OA_SOAP_SEN_BLADE_DISK_ZONE1, DISK_ZONE, 0},
+		{OA_SOAP_SEN_BLADE_MEM_ZONE1, MEMORY_ZONE, 1},
+		{OA_SOAP_SEN_TEMP_STATUS, AMBIENT_ZONE, 1}
+	},
 	/* BL480c blade type */
 	{
 		{OA_SOAP_SEN_BLADE_SYSTEM_ZONE1, SYSTEM_ZONE, 4},
@@ -10082,6 +10109,18 @@ const struct oa_soap_static_thermal_sens
 		{OA_SOAP_SEN_BLADE_MEM_ZONE1, MEMORY_ZONE, 1},
 		{OA_SOAP_SEN_TEMP_STATUS, AMBIENT_ZONE, 1}
 	},
+	/* BL685 G6 blade type */
+	{
+		{OA_SOAP_SEN_BLADE_SYSTEM_ZONE1, SYSTEM_ZONE, 3},
+		{OA_SOAP_SEN_BLADE_CPU_ZONE1, CPU_ZONE, 0},
+		{OA_SOAP_SEN_BLADE_CPU1_1, CPU_1, 1},
+		{OA_SOAP_SEN_BLADE_CPU2_1, CPU_2, 1},
+		{OA_SOAP_SEN_BLADE_CPU3_1, CPU_3, 1},
+		{OA_SOAP_SEN_BLADE_CPU4_1, CPU_4, 1},
+		{OA_SOAP_SEN_BLADE_DISK_ZONE1, DISK_ZONE, 0},
+		{OA_SOAP_SEN_BLADE_MEM_ZONE1, MEMORY_ZONE, 1},
+		{OA_SOAP_SEN_TEMP_STATUS, AMBIENT_ZONE, 1}
+	},
 	/* BL685 blade type */
 	{
 		{OA_SOAP_SEN_BLADE_SYSTEM_ZONE1, SYSTEM_ZONE, 2},
diff -up openhpi-2.14.1/plugins/oa_soap/oa_soap_resources.h.thermal-info-segfault openhpi-2.14.1/plugins/oa_soap/oa_soap_resources.h
--- openhpi-2.14.1/plugins/oa_soap/oa_soap_resources.h.thermal-info-segfault	2013-04-03 22:56:35.856017826 +0200
+++ openhpi-2.14.1/plugins/oa_soap/oa_soap_resources.h	2013-04-03 23:26:49.296892952 +0200
@@ -81,9 +81,11 @@ struct oa_soap_control {
 	const char *comment;
 };
 
-#define OA_SOAP_MAX_BLD_TYPE 19
+#define OA_SOAP_MAX_BLD_TYPE 22
 
-/* Enum for possible cclass blade types */
+/* Enum for possible cclass blade types 
+ * Put more specific ones always first
+ * */
 enum oa_soap_blade_type {
 	BL260C,
 	BL2x220C,
@@ -91,10 +93,13 @@ enum oa_soap_blade_type {
 	BL460CG8,
 	BL460CG7,
 	BL460C,
+	BL465CG7,
 	BL465C,
+	BL480CG1,
 	BL480C,
 	BL495C,
 	BL680C,
+	BL685CG6,
 	BL685C,
 	BL860C,
 	BL870C,
